pallets/price-oracle/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+
+mod benchmarking;
+
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+
+pub type AssetPairId = Vec<u8>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		PriceComputationError,
+		UpdateDataOverflow,
+	}
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
+
+	#[pallet::storage]
+	#[pallet::getter(fn price_buffer)]
+	pub type PriceBuffer<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, Vec<PriceEntry>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_ten)]
+	pub type PriceDataTen<T: Config> = StorageValue<_, Vec<(AssetPairId, BucketQueue)>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_hundred)]
+	pub type PriceDataHundred<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_thousand)]
+	pub type PriceDataThousand<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			let (num_of_processed_buckets, num_of_processed_trades) = Self::update_data().unwrap_or_else(|_| (0, 0));
+
+			PriceBuffer::<T>::remove_all(None);
+
+			T::WeightInfo::on_initialize_multiple_entries_multiple_tokens(num_of_processed_buckets, num_of_processed_trades) // TODO: rebenchmark
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn on_create_pool(asset_pair: AssetPair) {
+		let data = PriceDataTen::<T>::get();
+		if !data.iter().any(|bucket_tuple| bucket_tuple.0 == asset_pair.name()) {
+			PriceDataTen::<T>::append((asset_pair.name(), BucketQueue::default()));
+		}
+	}
+
+	pub fn on_trade(asset_pair: AssetPair, price_entry: PriceEntry) {
+		PriceBuffer::<T>::append(asset_pair.name(), price_entry);
+	}
+
+	fn update_data() -> Result<(u32, u32), DispatchError> {
+		let mut num_of_processed_buckets = 0u32;
+		let mut num_of_processed_trades = 0u32;
+
+		PriceDataTen::<T>::mutate(|data_ten| -> DispatchResult {
+			for (asset_pair_id, data) in data_ten.iter_mut() {
+				let maybe_price = <PriceBuffer<T>>::try_get(asset_pair_id);
+				let result = if let Ok(prices) = maybe_price {
+					num_of_processed_buckets.checked_add(1).ok_or(Error::<T>::UpdateDataOverflow)?;
+					num_of_processed_trades.checked_add(prices.len().try_into().map_err(|_| Error::<T>::UpdateDataOverflow)?).ok_or(Error::<T>::UpdateDataOverflow)?;
+					PriceInfo::calculate_price_info(prices.as_slice()).ok_or(Error::<T>::PriceComputationError)?
```

Comment:
This looks quadratic to me (number of asset pairs * number of prices per pair). You thus need to make sure to cover the max value of each in the benchmark to get a sensible linear estimate. (The linear regression run on the benchmark data thinks the number of asset pairs and the number of prices are independent, but they are not.)

Reply:
Hmm. If we store intermediate value after each trade with total weight we shouldn't need to recalculate it. Like so -> If you look at the equations you only need total weight and previous price to calculate current price instead of doing recalc of everything each time. https://docs.google.com/spreadsheets/d/1hYP8Y0qDbKOFSHcnP0Q3oyJmOwXTDLV79851Be5Hrgo/edit?usp=sharing

Reply:
@jak-pan it makes sense. It may help to solve the problem with on_trade() where it was complicated to calculate the weight because we don't know the number of already executed trades stored in the price buffer. I am going to test it.